In [57]:
import pandas as pd
import numpy as np

In [58]:
df=pd.read_csv(r"C:\Users\Admin\MSc Applied Statistics SPU\csv file\drug200.csv")

In [59]:
df.head()

Age Sex      BP Cholesterol  Na_to_K   Drug
0   23   F    HIGH        HIGH   25.355  drugY
1   47   M     LOW        HIGH   13.093  drugC
2   47   M     LOW        HIGH   10.114  drugC
3   28   F  NORMAL        HIGH    7.798  drugX
4   61   F     LOW        HIGH   18.043  drugY

In [60]:
df.shape

(200, 6)

In [61]:
df.columns

Index(['Age', 'Sex', 'BP', 'Cholesterol', 'Na_to_K', 'Drug'], dtype='object')

In [62]:
from sklearn.preprocessing import LabelEncoder,StandardScaler

In [65]:
df.head()

Age Sex      BP Cholesterol  Na_to_K   Drug
0   23   F    HIGH        HIGH   25.355  drugY
1   47   M     LOW        HIGH   13.093  drugC
2   47   M     LOW        HIGH   10.114  drugC
3   28   F  NORMAL        HIGH    7.798  drugX
4   61   F     LOW        HIGH   18.043  drugY

In [66]:
le={}
for i in df.select_dtypes(include='object').columns:
    le[i]=LabelEncoder()
    df[i]=le[i].fit_transform(df[i])

In [67]:
df.head()

Age  Sex  BP  Cholesterol  Na_to_K  Drug
0   23    0   0            0   25.355     4
1   47    1   1            0   13.093     2
2   47    1   1            0   10.114     2
3   28    0   2            0    7.798     3
4   61    0   1            0   18.043     4

In [68]:
X=df.drop("Drug",axis=1)
y=df["Drug"]

In [69]:
std=StandardScaler()
X_trns=std.fit_transform(X)

In [70]:
from sklearn.model_selection import train_test_split

In [71]:
X_train,X_test,y_train,y_test=train_test_split(X_trns,y,test_size=0.2,random_state=42)

In [72]:
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense

In [73]:
model = Sequential([
    Dense(32, activation='relu', input_shape=(X_train.shape[1],)),
    Dense(16, activation='relu'),
    Dense(5, activation='softmax')
])


In [74]:
model.compile(optimizer='adam',loss='sparse_categorical_crossentropy',metrics=['accuracy'])

In [75]:
model.fit(X_train,y_train,epochs=50,validation_split=0.1,batch_size=16)

Epoch 1/50
9/9 [==============================] - 3s 48ms/step - loss: 1.5613 - accuracy: 0.3542 - val_loss: 1.5713 - val_accuracy: 0.4375
Epoch 2/50
9/9 [==============================] - 0s 13ms/step - loss: 1.4948 - accuracy: 0.5208 - val_loss: 1.5005 - val_accuracy: 0.5625
Epoch 3/50
9/9 [==============================] - 0s 12ms/step - loss: 1.4307 - accuracy: 0.5417 - val_loss: 1.4260 - val_accuracy: 0.5625
Epoch 4/50
9/9 [==============================] - 0s 14ms/step - loss: 1.3638 - accuracy: 0.6042 - val_loss: 1.3506 - val_accuracy: 0.5625
Epoch 5/50
9/9 [==============================] - 0s 12ms/step - loss: 1.2965 - accuracy: 0.6319 - val_loss: 1.2766 - val_accuracy: 0.5625
Epoch 6/50
9/9 [==============================] - 0s 12ms/step - loss: 1.2264 - accuracy: 0.6597 - val_loss: 1.2000 - val_accuracy: 0.6250
Epoch 7/50
9/9 [==============================] - 0s 14ms/step - loss: 1.1584 - accuracy: 0.6597 - val_loss: 1.1286 - val_accuracy: 0.6875
Epoch 8/50
9/9 [===========

Setting **epochs=50** means the model will iterate over the entire training data 50 times to learn and adjust the weights.

In this case, the model will use 16 samples per batch. After processing 16 samples, the model will update the weights (i.e., the model will "learn" from those 16 samples).

Validation Data: The remaining $10%$ is used to evaluate the model's performance during each epoch (iteration) of training. It allows you to monitor the model's performance on unseen data, which helps you assess how well it generalizes. This helps in detecting issues like overfitting.

**Example:**
Suppose you have 1000 training samples:

With **validation_split=0.1** , 100 samples (10% of 1000) are set aside for validation.

The remaining 900 samples are used to train the model.

The model will train on those 900 samples and, after each epoch, will evaluate its performance using the 100 validation samples.

In [76]:
loss,accuracy=model.evaluate(X_test,y_test)
print(f"Test Accuracy: {accuracy * 100:.2f}%")

2/2 [==============================] - 0s 9ms/step - loss: 0.2225 - accuracy: 0.9500
Test Accuracy: 95.00%


The model is achieving good performance with a test accuracy of 95%. This means that, out of 100 test samples, the model is correctly predicting the class of 95 of them.

**Loss of 0.2225** indicates that the predictions are not perfect but relatively close to the true labels.

Depending on the complexity of the dataset and problem, **95%** accuracy can be considered very good. However, whether it's satisfactory depends on the specific context and the baseline performance you are comparing it to.

In [79]:
from sklearn.tree import DecisionTreeClassifier
from sklearn.metrics import accuracy_score

In [77]:
intermediate_layer_model = Sequential(model.layers[:-1])  # Remove last layer
X_transformed = intermediate_layer_model.predict(X_trns)

7/7 [==============================] - 0s 3ms/step


In [80]:
tree_model = DecisionTreeClassifier()
tree_model.fit(X_transformed, y)

DecisionTreeClassifier()

In [81]:
y_pred = tree_model.predict(X_transformed)
print("Hybrid Model Accuracy:", accuracy_score(y, y_pred))

Hybrid Model Accuracy: 1.0
